torch/_refs/__init__.py

```diff
@@ -1539,6 +1541,79 @@ def minimum(a: TensorLikeType, b: TensorLikeType) -> TensorLikeType:
     return prims.minimum(a, b)
 
 
+def _min_max_out(
+    result: TensorLikeType,
+    out: TensorLikeType,
+) -> TensorLikeType:
+    assert isinstance(out, TensorLike)
+    out = _maybe_resize_out(out, result.shape)
+    _safe_copy_out(copy_from=result, copy_to=out)  # type: ignore[arg-type]
+    return out
+
+
+def _min_max_out_tuple(
+    result: Tuple[TensorLikeType, TensorLikeType],
+    out: Tuple[TensorLikeType, TensorLikeType],
+) -> Tuple[TensorLikeType, TensorLikeType]:
+    rv, ri = result
+    ov, oi = out
+    assert isinstance(ov, TensorLike)
+    assert isinstance(oi, TensorLike)
+    ov = _maybe_resize_out(ov, rv.shape)
+    oi = _maybe_resize_out(oi, ri.shape)
+    _safe_copy_out(copy_from=rv, copy_to=ov)  # type: ignore[arg-type]
+    _safe_copy_out(copy_from=ri, copy_to=oi)  # type: ignore[arg-type]
+    return ov, oi
+
+
+@register_decomposition([aten.max.default, aten.max.dim])
+def max(
+    self: TensorLikeType,
+    other_or_dim: Union[TensorLikeType, int] = None,
+    keepdim: bool = False,
+    *,
+    out: Union[TensorLikeType, Tuple[TensorLikeType, TensorLikeType]] = None,
+) -> Union[TensorLikeType, Tuple[TensorLikeType, TensorLikeType]]:
+    if other_or_dim is None:
+        result = prims.max(self)
+        if out is None:
+            return result
+        # Cannot use out_wrapper because of the overloads
+        return _min_max_out(result, out)
+    elif isinstance(other_or_dim, Tensor):
+        return torch.maximum(self, other_or_dim, out=out)
+    assert isinstance(other_or_dim, Dim)
+    result = prims.max_dim(self, other_or_dim, keepdim=keepdim)
+    if out is None:
+        return result
+    # Cannot use out_wrapper because of the overloads
+    return _min_max_out_tuple(result, out)
```

Comment:
This handles all overloads together, like `mean`:

```
 * (Tensor input, *, Tensor out)
 * (Tensor input, Tensor other, *, Tensor out)
 * (Tensor input, int dim, bool keepdim, *, tuple of Tensors out)
 * (Tensor input, name dim, bool keepdim, *, tuple of Tensors out)  # this one is ignored because it uses named dims
```